src/material/datepicker/datepicker-actions.ts

```diff
@@ -0,0 +1,87 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  AfterViewInit,
+  ChangeDetectionStrategy,
+  Component,
+  Directive,
+  OnDestroy,
+  TemplateRef,
+  ViewChild,
+  ViewContainerRef,
+  ViewEncapsulation
+} from '@angular/core';
+import {TemplatePortal} from '@angular/cdk/portal';
+import {MatDatepickerBase, MatDatepickerControl} from './datepicker-base';
+
+
+/** Button that will close the datepicker and assign the current selection to the data model. */
+@Directive({
+  selector: '[matDatepickerApply], [matDateRangePickerApply]',
+  host: {'(click)': '_applySelection()'}
+})
+export class MatDatepickerApply {
+  constructor(private _datepicker: MatDatepickerBase<MatDatepickerControl<unknown>, unknown>) {}
+
+  _applySelection() {
+    this._datepicker._applyPendingSelection();
+    this._datepicker.close();
+  }
+}
+
+
+/** Button that will close the datepicker and discard the current selection. */
+@Directive({
+  selector: '[matDatepickerCancel], [matDateRangePickerCancel]',
+  host: {'(click)': '_datepicker.close()'}
+})
+export class MatDatepickerCancel {
+  constructor(public _datepicker: MatDatepickerBase<MatDatepickerControl<unknown>, unknown>) {}
+}
+
+
+/**
+ * Container that can be used to project a row of action buttons
+ * to the bottom of a datepicker or date range picker.
+ */
+@Component({
+  selector: 'mat-datepicker-actions, mat-date-range-picker-actions',
+  styleUrls: ['datepicker-actions.css'],
+  template: `
+    <ng-template>
```

Comment:
Is there a way we could avoid doing the `ng-content` inside `ng-template` pattern? Doing this, Angular ends up instantiating the content even when it's not rendered. It could just be an `ng-template`-based API, or maybe just vanilla content projection?

Reply:
We could, but I decided not to do it, because projecting a couple of buttons isn't particularly performance-intensive. Also using an `ng-template` to define the buttons doesn't allow for things like `ngIf` and `ngFor` to be used on the same element.